bigframes/pandas/__init__.py

```diff
@@ -134,6 +138,136 @@ def cut(
+        if pandas.api.types.is_list_like(kwarg) and len(kwarg) != len(columns):
+            raise ValueError(
+                f"Length of '{kwarg_name}' ({len(kwarg)}) did not match "
+                f"the length of the columns being encoded ({len(columns)})."
+            )
+        elif pandas.api.types.is_list_like(kwarg):
+            return list(map(str, kwarg))
+        elif isinstance(kwarg, str):
+            return [kwarg] * len(columns)
+        elif isinstance(kwarg, dict):
+            return [kwarg[column] for column in columns]
+        elif kwarg is None:
+            return None
+        else:
+            raise TypeError(f"{kwarg_name} kwarg must be a string, list, or dictionary")
+
+    prefix_sep = parse_prefix_kwarg(prefix_sep, "prefix_sep")
+    if prefix_sep is None:
+        prefix_sep = ["_"] * len(columns)
+    prefix = parse_prefix_kwarg(prefix, "prefix")
+    if prefix is None:
+        prefix = columns
+
+    max_unique_value = (
+        bigframes.core.blocks._BQ_MAX_COLUMNS
+        - len(block.value_columns)
+        - len(block.index_columns)
+        - 1
+    ) // len(columns)
+    columns_ids = []
+    full_prefixes_with_duplicity = []
+    for i in range(len(columns)):
+        label = columns[i]
+        full_prefix = prefix[i] + prefix_sep[i]
+        for col_id in block.label_to_col_id[label]:
+            columns_ids.append(col_id)
+            full_prefixes_with_duplicity.append(full_prefix)
+
+    columns_values = block._get_unique_values(columns_ids, max_unique_value)
+
+    result = block.drop_columns(columns_ids)
+    for i in range(columns_values.nlevels):
+        level = columns_values.get_level_values(i).sort_values()
+        column_label = full_prefixes_with_duplicity[i]
+        column_id = columns_ids[i]
+
+        already_seen = set()
+        for value in level:
+            if pandas.isna(value):
+                continue
+            if value in already_seen:
+                continue
+            already_seen.add(value)
+            if drop_first and len(already_seen) == 1:
+                continue
+
+            new_column_label = f"{column_label}{value}"
+            if isinstance(data, Series) and not prefix_given:
+                # in this special case, the columns are labeled with
+                # the scalar itself, not a string
+                new_column_label = value
+
+            new_block, new_id = block.apply_unary_op(
+                column_id, ops.BinopPartialLeft(ops.eq_op, value)
+            )
+            new_block, new_id = new_block.apply_unary_op(
+                new_id, ops.BinopPartialRight(ops.fillna_op, False)
+            )
+
+            new_block = new_block.select_column(new_id)
+            new_block = new_block.with_column_labels([new_column_label])
+            result_index, _ = result.index.join(
+                new_block.index, how="left", block_identity_join=True
+            )
+            result = result_index._block
```

Comment:
might be better just to select all the ids at once at the end instead of relying so heavily on the implicit joiner to avoid doing a 'real' join

Reply:
Got it, apply_unary_op already does what I want it to do (add just 1 column to the block), so I can avoid any sort of join call here.